src/schema_compile.c

```diff
@@ -1231,27 +1271,15 @@ lys_compile_unres_depset(struct ly_ctx *ctx, struct lys_glob_unres *unres)
      * can be also leafref, in case it is already resolved, go through the chain and check that it does not
      * point to the starting leafref type). The second round stores the first non-leafref type for later data validation. */
     for (i = 0; i < ds_unres->leafrefs.count; ++i) {
-        LY_ERR ret = LY_SUCCESS;
         node = ds_unres->leafrefs.objs[i];
         cctx.cur_mod = node->module;
         cctx.pmod = node->module->parsed;
-
         LOG_LOCSET(node, NULL, NULL, NULL);
 
         assert(node->nodetype & (LYS_LEAF | LYS_LEAFLIST));
-        type = ((struct lysc_node_leaf *)node)->type;
-        if (type->basetype == LY_TYPE_LEAFREF) {
-            ret = lys_compile_unres_leafref(&cctx, node, (struct lysc_type_leafref *)type, unres);
-        } else if (type->basetype == LY_TYPE_UNION) {
-            LY_ARRAY_FOR(((struct lysc_type_union *)type)->types, v) {
-                if (((struct lysc_type_union *)type)->types[v]->basetype == LY_TYPE_LEAFREF) {
-                    lref = (struct lysc_type_leafref *)((struct lysc_type_union *)type)->types[v];
-                    ret = lys_compile_unres_leafref(&cctx, node, lref, unres);
-                    if (ret) {
-                        break;
-                    }
-                }
-            }
+        v = 0;
+        while ((ret == LY_SUCCESS) && (lref = lys_type_leafref_next(node, &v))) {
```

Comment:
I would personally keep this condition only for non-NULL `lref` and `break` in case the function in the loop returns an error. But you are free to keep it this way if you think it is more readable.

Reply:
I think it's more readable because the code is shorter and the condition only consists of two parts, so when debugging it shouldn't be such a problem to check why the loop ended.

Reply:
Why I prefer the other solution is more because I think the conditions are different enough not to be in a single expression. In other words, I like error conditions to always be handled separately. But no real problem, just something to think about.